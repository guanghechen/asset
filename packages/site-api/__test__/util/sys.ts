/**
 * Create a promise resolved after ${milliseconds} ms.
 *
 * @param milliseconds
 */
export function noop(milliseconds: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, milliseconds)
  })
}
