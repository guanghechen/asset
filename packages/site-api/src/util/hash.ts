import crypto from 'crypto'

/**
 * Calc hash value with sha1 algorithm
 *
 * @param content
 */
export function sha1(content: string | Buffer): string {
  const sha1 = crypto.createHash('sha1')
  sha1.update(content)
  return sha1.digest('hex')
}


/**
 * calc fingerprint of content
 * @param content
 */
export function calcFingerprint(content: Buffer): string {
  return sha1(content)
}


/**
 * Generate unique id from text
 *
 * @param text
 */
export function uniqueText(
  text: string | null,
): typeof text extends null ? null : string {
  if (text == null) return null as any
  return text.toLowerCase().trim().replace(/[\s\n]+/g, '--')
}
