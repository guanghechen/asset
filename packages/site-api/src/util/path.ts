import path from 'path'


/**
 * Resolve url path
 *
 * @param prefixPath
 * @param p
 */
export function resolveUrlPath(
  prefixPath: string,
  ...pathPieces: string[]
): string {
  const result = (prefixPath + '/' + pathPieces.join('/'))
    .replace(/[\\/]+/g, '/')
    .replace(/^[/]?/, '/')
    .replace(/[/]$/, '')
  return result
}


/**
 * Resolve local data path
 *
 * @param basePath
 * @param p
 */
export function resolveLocalPath(
  basePath: string,
  ...pathPieces: string[]
): string {
  const result = path.resolve(basePath, ...pathPieces)
  return result
}


/**
 * Calculate filepath cross platform and repository
 *
 * @param workspace
 * @param filepath
 */
export function resolveUniversalPath(
  workspace: string,
  filepath: string,
): string {
  const result = path
    .normalize(path.relative(workspace, path.resolve(workspace, filepath)))
    .replace(/[\\/]+/g, '/')
    .replace(/[/]$/, '')
  return result
}
