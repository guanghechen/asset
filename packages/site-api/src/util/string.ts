/**
 * Stringify data
 * @param data
 */
export function stringify(data: unknown): string {
  return JSON.stringify(data, null, 2)
}
