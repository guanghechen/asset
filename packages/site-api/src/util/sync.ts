/**
 * Async task
 */
export interface AsyncTask<D extends unknown = unknown> {
  /**
   * Task data
   */
  data: D
  /**
   * Execute task
   */
  execute: () => (void | Promise<void>)
}


/**
 * Executor to execute asynchronous tasks serially
 */
export interface SerialExecutor<D extends unknown = unknown> {
  /**
   * Append task to executor
   */
  addTask: (task: AsyncTask<D>) => void
}


/**
 * Create a SerialExecutor to execute asynchronous tasks serially
 *
 * @param squashable      check if current task could be skipped
 * @param onTaskSuccess   triggered on each task finished
 * @param onTaskFailure   triggered on each task crashed
 * @param onTaskCompleted triggered on each task terminated (finished / crashed)
 */
export function createSerialExecutor<D extends unknown = unknown>(
  squashable?: (currentData: D, nextData: D) => boolean,
  onTaskSuccess?: () => void | Promise<void>,
  onTaskFailure?: (error: any) => void | Promise<void>,
  onTaskCompleted?: (error?: any) => void | Promise<void>,
): SerialExecutor<D> {
  let running = false
  const tasks: AsyncTask<D>[] = []

  const runTask = async () => {
    // If running or no task remain, no operation will be performed
    if (running || tasks.length <= 0) return

    // Ready to start task
    running = true
    const task = tasks.shift()!

    // Optimization: check if current task could be skipped
    if (tasks.length > 0) {
      const nextTask = tasks[0]
      if (squashable != null && squashable(task.data, nextTask.data)) {
        running = false
        await runTask()
        return
      }
    }

    let error: any | undefined
    try {
      await task.execute()
    } catch (err) {
      error = err
    } finally {
      // Run hooks
      if (error != null) {
        if (onTaskFailure) {
          await onTaskFailure(error)
        } else {
          throw error
        }
      } else {
        if (onTaskSuccess) await onTaskSuccess()
      }
      if (onTaskCompleted) await onTaskCompleted(error)

      // Release lock
      running = false

      // Start next task
      if (tasks.length > 0) await runTask()
    }
  }

  const addTask = (task: AsyncTask<D>) => {
    tasks.push(task)
    runTask()
  }

  return { addTask }
}
