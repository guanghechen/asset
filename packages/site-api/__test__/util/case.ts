import path from 'path'
import {
  SiteConfig,
  createDefaultSubSiteConfig,
  loadSiteConfig,
  resolveSubSiteConfig,
  resolveSubSiteSourceItem,
} from '../../src'

const caseRootDir = path.resolve(__dirname, '../cases')
const resolveCasePath = (...relativePieces: string[]): string => (
  path.resolve(caseRootDir, ...relativePieces))


export interface CaseItem {
  /**
   * Case title
   */
  title: string
  /**
   * Root dir of the case
   */
  workspace: string
  /**
   * Site config provider
   */
  getSiteConfig: () => SiteConfig
}


export const cases: CaseItem[] = [
  // load full config
  ((): CaseItem => {
    const title = 'full-config'
    const workspace = resolveCasePath(title)
    const configFilepath = path.resolve(workspace, '_config.yml')

    const getSiteConfig = (): SiteConfig => {
      return loadSiteConfig(
        workspace,
        configFilepath,
        undefined,
        undefined,
        {
          barusu: function (rawConfig, sitePathConfig, defaultConfig) {
            return resolveSubSiteConfig(
              rawConfig,
              ['image'],
              resolveSubSiteSourceItem,
              sitePathConfig,
              {
                ...createDefaultSubSiteConfig('barusu'),
                source: {
                  image: {
                    sourceRoot: 'asset/image',
                    dataRoot: 'asset/image',
                    pattern: ['**/*{.png,jpg}'],
                  }
                },
              },
            )
          }
        }
      )
    }

    return { title, workspace, getSiteConfig }
  })()
]
