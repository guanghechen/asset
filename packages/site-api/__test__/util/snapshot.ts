import path from 'path'
import stripAnsi from 'strip-ansi'

const WORKSPACE_ROOT_DIR = path.resolve(__dirname, '../..')


/**
 *
 * @param key
 * @param value
 */
export const assetDataReplacer = (key: string, value: unknown): unknown | undefined => {
  if (['createAt', 'lastModifiedTime', 'updateAt'].includes(key)) {
    return '<' + typeof value + '>'
  }
  return typeof value === 'string' ? stripAnsi(value) : value
}


/**
 * Remove sensitive data
 *
 * @param data
 */
export function desensitize<D extends unknown = unknown>(
  data: D,
  replacer?: (key: string, value: unknown) => unknown | undefined,
): D {
  // No replaceAll, so use .split().join as a alternative
  const str = JSON.stringify(data, replacer)
    .split(WORKSPACE_ROOT_DIR)
    .join('<PACKAGE_ROOT_DIR>')
  return JSON.parse(str)
}
