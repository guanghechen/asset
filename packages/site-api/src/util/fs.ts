import fs from 'fs-extra'
import path from 'path'

/**
 * Ensure the dirpath exists
 * @param dirpath
 */
export function ensurePathExists(dirpath: string): void {
  if (!fs.existsSync(dirpath)) {
    fs.mkdirsSync(dirpath)
  }
}

/**
 * Output into file
 *
 * @param filepath
 * @param data
 */
export async function writeFile(filepath: string, data: Buffer): Promise<void> {
  ensurePathExists(path.dirname(filepath))
  await fs.writeFile(filepath, data)
}

/**
 * Output into file (synchronous)
 *
 * @param filepath
 * @param data
 */
export function writeFileSync(filepath: string, data: Buffer): void {
  ensurePathExists(path.dirname(filepath))
  fs.writeFileSync(filepath, data)
}

/**
 * Output into json file
 *
 * @param filepath
 * @param data
 */
export async function writeJSON<T extends unknown = unknown>(
  filepath: string,
  data: T,
): Promise<T> {
  ensurePathExists(path.dirname(filepath))
  await fs.writeJSON(filepath, data)
  return data
}

/**
 * Output into json file (synchronous)
 *
 * @param filepath
 * @param data
 */
export function writeJSONSync<T extends unknown = unknown>(
  filepath: string,
  data: T,
): T {
  ensurePathExists(path.dirname(filepath))
  fs.writeJSONSync(filepath, data)
  return data
}
