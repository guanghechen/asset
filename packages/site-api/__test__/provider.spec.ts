import fs from 'fs-extra'
import path from 'path'
import {
  AssetDataItem,
  AssetDataProvider,
  AssetEntity,
  AssetEntityManager,
  AssetProcessor,
  CategoryDataItem,
  TagDataItem,
} from '@guanghechen/site-api'
import { cases } from './util/case'
import { assetDataReplacer, desensitize } from './util/snapshot'
import { noop } from './util/sys'


describe('AssetDataProvider', function () {
  jest
    .spyOn(global.console, 'info')
    .mockImplementation(function (...args: string[]): any {
      expect(desensitize(args, assetDataReplacer))
        .toMatchSnapshot('console.info')
    })

  for (const kase of cases) {
    describe(kase.title, function () {
      const { getSiteConfig } = kase
      const siteConfig = getSiteConfig()
      const fsWriteJsonSyncMock = jest
        .spyOn(fs, 'writeJSON')
        .mockImplementation(function (
          absolutePath: string | any,
          data: any,
        ): any {
          const snapshotName = path.relative(siteConfig.workspace, absolutePath)
          expect(desensitize(data, assetDataReplacer))
            .toMatchSnapshot(desensitize(snapshotName))
        })

      const subSiteNames = Object.keys(siteConfig.sites)
      for (const subSiteName of subSiteNames) {
        describe(subSiteName, function () {
          const subSiteConfig = siteConfig.sites[subSiteName]

          const fakeEntityManagerMap: Record<string, AssetEntityManager<AssetEntity>> = {}
          const fakeProcessor: AssetProcessor = {
            processable: filepath => /\.([^.]+)\.json$/.test(filepath),
            process: (filepath, rawContent, roughAsset, tagDataManager, categoryDataManager) => {
              const data = JSON.parse(rawContent.toString('utf-8'))

              const tags: TagDataItem[] = (data.tags || []).map((rawTag: string) => (
                tagDataManager.normalize(rawTag)
              ))

              const categories: CategoryDataItem[][] = (data.categories || []).map(
                (categoryPath: string[]) => (
                  categoryPath.map(c => categoryDataManager.normalize(c))))

              const asset: AssetEntity = {
                ...roughAsset,
                type: /\.([^.]+)\.json$/.exec(filepath)![1],
                tags: tags.map(tag => tag.uuid),
                categories: categories.map(cp => cp.map(c => c.uuid)),
              }

              if (fakeEntityManagerMap[asset.type] == null) {
                fakeEntityManagerMap[asset.type] = new AssetEntityManager(
                  subSiteConfig.source[asset.type].dataRoot)
              }

              const entityManager = fakeEntityManagerMap[asset.type]
              entityManager.insert(asset)
              return [asset, tags, categories]
            }
          }

          const getCurrentData = (provider: AssetDataProvider<any>): unknown[] => {
            const assets = Object.keys(subSiteConfig.source)
              .reduce((acc, assetType) => {
                const assets = provider.assetService.fetchAssets(assetType)
                acc.push(...assets)
                return acc
              }, [] as any[])
            const tags = provider.tagService.fetchTags()
            const categories = provider.categoryService.fetchCategories()
            return [assets, tags, categories]
          }

          const baseTest = function (provider: AssetDataProvider<any>) {
            // test tags
            const tags: TagDataItem[] = provider.tagService.fetchTags()
            expect(tags.map(t => provider.tagService.fetchTag(t.uuid))).toEqual(tags)
            expect(provider.tagService.fetchTag('non-exist uuid')).toBeNull()

            // test categories
            const categories: CategoryDataItem[] = provider.categoryService.fetchCategories()
            expect(
              categories.map(c => provider.categoryService.fetchCategory(c.uuid))
            ).toEqual(categories)
            expect(provider.categoryService.fetchCategory('non-exist uuid')).toBeNull()

            // test assets
            for (const assetType of Object.keys(subSiteConfig.source)) {
              const assets: AssetDataItem[] = provider.assetService.fetchAssets(assetType)
              expect(assets.map(a => provider.assetService.fetchAsset(a.uuid))).toEqual(assets)
              expect(provider.assetService.fetchAsset('non-exist uuid')).toBeNull()

              const entityManager = fakeEntityManagerMap[assetType]
              if (entityManager == null) {
                expect(assets.length).toBeLessThanOrEqual(0)
              } else {
                expect(entityManager.find('non-exist uuid')).toBeNull()
              }

              expect(
                provider.assetService.fetchAssets(assetType, 0, assets.length)
              ).toEqual(assets)
              expect(
                provider.assetService.fetchAssets(assetType, 0, assets.length - 1)
              ).toEqual(assets.slice(0, -1))

              // test asset entity
              expect(
                assets.map(a => {
                  const entity = entityManager.find(a.uuid)!
                  const result = {}
                  for (const key of Object.keys(a)) {
                    result[key] = entity[key]
                  }
                  return result
                })
              ).toEqual(assets)
            }
          }

          test('build', async function () {
            const provider = new AssetDataProvider(subSiteConfig, [fakeProcessor])
            await provider.build(true)

            // test tags
            const tags: TagDataItem[] = provider.tagService.fetchTags()
            expect(desensitize(tags, assetDataReplacer)).toMatchSnapshot('tags')

            // test categories
            const categories: CategoryDataItem[] = provider.categoryService.fetchCategories()
            expect(desensitize(categories, assetDataReplacer)).toMatchSnapshot('categories')

            // test assets
            for (const assetType of Object.keys(subSiteConfig.source)) {
              const assets: AssetDataItem[] = provider.assetService.fetchAssets(assetType)
              expect(desensitize(assets, assetDataReplacer)).toMatchSnapshot('asset -- ' + assetType)
            }

            baseTest(provider)
          })

          test('start', async function () {
            const provider = new AssetDataProvider(subSiteConfig, [fakeProcessor])
            await provider.watch(true)
            baseTest(provider)
            const initialData = getCurrentData(provider)

            const testImageData = {
              tags: ['gif', 'jpg', 'jpeg'],
              categories: [
                ['test', 'provider', 'image'],
                ['asset', 'image'],
              ]
            }
            const testImageFilepath = path.resolve(
              subSiteConfig.source.image.sourceRoot, 'provider.test.image.json')

            await fs.writeFile(testImageFilepath, JSON.stringify(testImageData), 'utf-8')
            await noop(2000)
            baseTest(provider)

            // test tags
            expect(
              provider.tagService.fetchTags()
              .filter(tag => testImageData.tags.includes(tag.title))
              .map(tag => tag.title)
              .sort()
            ).toEqual([...testImageData.tags].sort())

            // test categories
            expect(
              provider.categoryService.fetchCategories()
              .filter(category => testImageData.categories.flat().includes(category.title))
              .map(category => category.title)
              .sort()
            ).toEqual([...new Set(testImageData.categories.flat())].sort())

            await fs.unlink(testImageFilepath)
            await noop(2000)
            baseTest(provider)

            const finalData = getCurrentData(provider)
            expect(finalData).toEqual(initialData)

            await provider.close()
          })
        })
      }

      fsWriteJsonSyncMock.mockRestore()
    })
  }
})
