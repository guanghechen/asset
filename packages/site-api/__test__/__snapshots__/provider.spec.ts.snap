// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetDataProvider full-config barusu build: asset -- image 1`] = `
Array [
  Object {
    "categories": Array [
      Array [
        "math",
        "react--context",
        "waw",
      ],
      Array [
        "200",
        "react--context",
      ],
    ],
    "createAt": "<string>",
    "fingerprint": "d85cbb58e6fad6d146139a90742163afeeedffee",
    "lastModifiedTime": "<number>",
    "location": "a.image.json",
    "tags": Array [
      "png",
      "react--image--component",
    ],
    "title": "a.image",
    "type": "image",
    "updateAt": "<string>",
    "uuid": "643297f8cb62042d714240ec6cb2190ee8586873",
  },
]
`;

exports[`AssetDataProvider full-config barusu build: categories 1`] = `
Array [
  Object {
    "assets": Array [],
    "children": Array [
      "react--context",
    ],
    "parents": Array [],
    "title": "math",
    "uuid": "math",
  },
  Object {
    "assets": Array [
      "643297f8cb62042d714240ec6cb2190ee8586873",
    ],
    "children": Array [
      "waw",
    ],
    "parents": Array [
      "math",
      "200",
    ],
    "title": "React Context",
    "uuid": "react--context",
  },
  Object {
    "assets": Array [
      "643297f8cb62042d714240ec6cb2190ee8586873",
    ],
    "children": Array [],
    "parents": Array [
      "react--context",
    ],
    "title": "waw",
    "uuid": "waw",
  },
  Object {
    "assets": Array [],
    "children": Array [
      "react--context",
    ],
    "parents": Array [],
    "title": "200",
    "uuid": "200",
  },
]
`;

exports[`AssetDataProvider full-config barusu build: console.info 1`] = `
Array [
  "clearing <PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data",
]
`;

exports[`AssetDataProvider full-config barusu build: tags 1`] = `
Array [
  Object {
    "assets": Array [
      "643297f8cb62042d714240ec6cb2190ee8586873",
    ],
    "title": "png",
    "uuid": "png",
  },
  Object {
    "assets": Array [
      "643297f8cb62042d714240ec6cb2190ee8586873",
    ],
    "title": "React image component",
    "uuid": "react--image--component",
  },
]
`;

exports[`AssetDataProvider full-config barusu start: console.info 1`] = `
Array [
  "rebuilding <PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data",
]
`;

exports[`AssetDataProvider full-config barusu start: console.info 2`] = `
Array [
  "clearing <PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data",
]
`;

exports[`AssetDataProvider full-config barusu start: console.info 3`] = `
Array [
  "loading <PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data",
]
`;

exports[`AssetDataProvider full-config barusu start: console.info 4`] = `
Array [
  "watching <PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/source",
]
`;

exports[`AssetDataProvider full-config barusu start: console.info 5`] = `
Array [
  "closing watch mode",
]
`;
